### PR TITLE
HTTP/2 close only send GO_AWAY if one has not already been sent

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Http2ConnectionHandler will always send a GO_AWAY when the channel is closed. This may cause problems if the user is attempting to control when GO_AWAY is sent and the content of the GO_AWAY.

Modifications:
- When the channel is closed Http2ConnectionHandler should only send a GO_AWAY if one has not already been sent

Result:
The user has more control over when GO_AWAY is sent
Fixes https://github.com/netty/netty/issues/5307

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler @ejona86 - PTAL

**Comment by normanmaurer** [comment]
LGTM

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 1]
```diff
@@ -417,16 +417,28 @@ public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exce
...
 
     private void doGracefulShutdown(ChannelHandlerContext ctx, ChannelFuture future, ChannelPromise promise) {
         // If there are no active streams, close immediately after the send is complete.
         // Otherwise wait until all streams are inactive.
         if (isGracefulShutdownComplete()) {
-            future.addListener(new ClosingChannelFutureListener(ctx, promise));
+            if (future == null) {
+                ctx.close(promise);
+            } else {
+                future.addListener(new ClosingChannelFutureListener(ctx, promise));
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Should the comment above be moved this this `else`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 2]
```diff
@@ -417,16 +417,28 @@ public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exce
...
         doGracefulShutdown(ctx, future, promise);
     }
 
     private void doGracefulShutdown(ChannelHandlerContext ctx, ChannelFuture future, ChannelPromise promise) {
         // If there are no active streams, close immediately after the send is complete.
         // Otherwise wait until all streams are inactive.
         if (isGracefulShutdownComplete()) {
-            future.addListener(new ClosingChannelFutureListener(ctx, promise));
+            if (future == null) {
+                ctx.close(promise);
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Maybe add a comment here to describe what you're doing?

**Comment by nmittler** [comment]
@Scottmitch LGTM

@ejona86 do you want to take a look?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I will clarify the comments.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I will clarify the comments.

**Comment by Scottmitch** [comment]
@nmittler - I changed the code slightly. Instead of not sending anything and immediately closing the connection if a go_away had already been sent ... I am now writing an empty buffer so I can wait until all pending writes are flushed before closing. This way if the user had written a GO_AWAY but not yet flushed ... we will force a flush and wait for the data to be written to the OS.

**Comment by nmittler** [comment]
@Scottmitch ah good catch ... sgtm

**Comment by nmittler** [comment]
@Scottmitch still LGTM

**Comment by normanmaurer** [comment]
LGTM...

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 3]
```diff
@@ -16,6 +16,8 @@
 
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.EmptyByteBuf;
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: MINOR Remove this unused import 'io.netty.buffer.EmptyByteBuf'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 4]
```diff
@@ -16,6 +16,8 @@
...
+import io.netty.buffer.Unpooled;
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: MINOR Remove this unused import 'io.netty.buffer.Unpooled'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by normanmaurer** [comment]
@Scottmitch please address netkins comments and then cherry-pick FTW :)

**Comment by Scottmitch** [comment]
@normanmaurer - Issues addressed.

Cherry-picked 4.1 (79f2e36)

**Comment by ejona86** [comment]
Late, but LGTM

**Comment by jchambers** [comment]
@Scottmitch I'm pretty sure I know why you're fixing this—thanks ;)

**Pull request closed without merging**